### PR TITLE
T/42: Various issues when dropping or inserting files. SRP in the command

### LINKED ISSUE #29: If the editor has focus, drop location is not read correctly
It may be more an issue of the clipboard or engine themselves, though.

may-10-2017 15-43-32

### LINKED ISSUE #42: The image should replace the current block if it's empty
jul-05-2017 19-23-39

### PR DESCRIPTION
**Pull request by Reinmar**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Fix: Fixed two issues related to dropping images. First, when dropping a file into an empty paragraph, that paragraph should be replaced with that image. Second, drop position should be read correctly when the editor is focused upon drop. Closes #42. Closes #29.

BREAKING CHANGE: `UploadImageCommand` doesn't optimize the drop position itself anymore. Instead, a separate `findOptimalInsertionPosition()` function was introduced.

BREAKING CHANGE: `UploadImageCommand` doesn't verify the type of file anymore. This needs to be done by the caller.

---

### Additional information

I started with minimal changes,  but it soon turned out that they lead to a lot more changes. I added some missing tests and it's possible that I actually fixed a few more bugs, but I didn't check if those things didn't work in the past.

### CODE REVIEW

File: src/utils.js [hunk 1]
```diff
@@ -19,3 +21,47 @@ export function isImageType( file ) {
 	return types.test( file.type );
 }
 
+/**
+ * Returns a model position which is optimal (in terms of UX) for inserting and image.
```
Comment (pjasiun) [inline src/utils.js]: for inserting and image?

Comment (Reinmar) [inline src/utils.js]: an image ofc

File: src/utils.js [hunk 2]
```diff
@@ -19,3 +21,47 @@ export function isImageType( file ) {
...
+ *
+ * For instance, if a selection is in a middle of a paragraph, position before this paragraph
```
Comment (pjasiun) [inline src/utils.js]: "For instance", and then you described all cases ;)

File: src/utils.js [hunk 3]
```diff
@@ -19,3 +21,47 @@ export function isImageType( file ) {
...
+ *
+ * @param {module:engine/model/selection~Selection} selection Selection based on which the
+ * insertion position should be calculated.
+ * @returns {module:engine/model/position~Position} The optimal position.
+ */
+export function findOptimalInsertionPosition( selection ) {
+	const selectedElement = selection.getSelectedElement();
+
+	if ( selectedElement ) {
+		return ModelPosition.createAfter( selectedElement );
```
Comment (pjasiun) [inline src/utils.js]: Comment what case it is?

**Comment by pjasiun** [comment]
The code looks and works great. After you update docs and comments you can merge this PR.

File: src/utils.js (hunk 3 above)
Comment (Reinmar) [inline src/utils.js]: The code reads so easily that I don't know what else you can write. If there's some element selected, return position after this selected element. It's nearly 1:1 to natural language :D

File: src/utils.js (hunk 2 above)
Comment (Reinmar) [inline src/utils.js]: No, I didn't. Element selection, edge cases like no blocks, etc.

**Pull request merged**